python-runner/runner.py

```diff
@@ -192,40 +315,57 @@ def apply_pipeline(record: dict, pipeline: dict, preview_step=None):
                                 )
                 record["value"] = value
 
+            # Initialize metadata keeping track of applied changes
+            if record["metadata"].get("lastChange") is None:
+                record["metadata"]["lastChange"] = {"key": {}, "value": {}}
+            # Check which fields of the key have been changed by this step
+            if type(record["key"]) is dict:
+                for key in record["key"].keys():
+                    if record["key"].get(key) != old_record["key"].get(key):
+                        record["metadata"]["lastChange"]["key"][key] = step_index
+            # Check which fields of the value have been changed by this step
+            if type(record["value"]) is dict:
+                for key in record["value"].keys():
+                    if record["value"].get(key) != old_record["value"].get(key):
+                        record["metadata"]["lastChange"]["value"][key] = step_index
+
             if preview_step == step_index:
                 return record
 
             step_index = step_index + 1
         return record
     except:  # noqa: E722
-        error_type = sys.exc_info()[0]
-        file_name = None
-        line_number = None
-        formatted_error = traceback.format_exc()
-        formatted_errors = formatted_error.split("\n")
-        if len(formatted_errors) > 3:
-            file_line_error = formatted_errors[-4].strip()
-            if match := re.search('File "(.*)".*', file_line_error):
-                file_name = match.group(1)
-            if match := re.search(".*line ([0-9]*).*", file_line_error):
-                line_number = int(match.group(1))
-        error_message = None
-        if len(formatted_errors) > 1:
-            error_message = formatted_errors[-2]
-
-        # inject error information into metadata
-        record["metadata"]["error"] = {
-            "location": location,
-            "exceptionMessage": error_message,
-            "exceptionName": error_type.__name__,
-            "file": {
-                "name": file_name,
-                "lineNumber": line_number,
-            },
-            "stacktrace": formatted_errors,
-        }
-
-        return record
+        return add_error_to_metadata(record, location, sys.exc_info()[0], traceback)
+
+
+def add_error_to_metadata(record, location, error_type, traceback):
```

Comment:
Would be ok to have this in this PR. But we should consider visualising errors on a field level. Not sure if this is possible this way.

Reply:
Correcting my previous comment. We should be able to visualise errors on a cell level.

Reply:
This is still possible since this metadata belongs to a specific record.